### ISSUE TITLE
migrate from kmsgrab

### ISSUE THREAD

**Issue by ReenigneArcher** [LizardByte/Sunshine#3327]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your issue described in the documentation?

- [X] I have read the documentation

### Is your issue present in the latest beta/pre-release?

This issue is present in the latest pre-release

### Describe the Bug

kmsgrab is deprecated/unsupported

### Additional Context

- https://github.com/LizardByte/Sunshine/issues/3298#issuecomment-2434431722
- https://github.com/LizardByte/Sunshine/issues/2955#issuecomment-2285259062
- https://github.com/LizardByte/Sunshine/issues/2250#issuecomment-2192072235

### Host Operating System

Linux

### Operating System Version

*

### Architecture

other, n/a

### Sunshine commit or version

*

### Package

other (not listed)

### GPU Type

n/a

### GPU Model

*

### GPU Driver/Mesa Version

*

### Capture Method

KMX (Linux)

### Relevant log output

```shell
*
```

**Comment by Zamundaaa** [comment]
FWIW one issue you'll come across with this is that the screencast portal doesn't support HDR streaming yet. I'd be happy to help out with fixing that on the KWin side though

**Comment by Quackdoc** [comment]
I believe there are two primary options for this

* Pipewire capture support + portals
* The new wayland protocols https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/124 

the wayland protocol is an alternative to wlr-screencapture and has PRs for both cosmic and wlroots. https://gitlab.freedesktop.org/wlroots/wlroots/-/merge_requests/4545 https://github.com/pop-os/cosmic-comp/pull/365

I would love to see both supported. Pipewire + xdg portals gives a good basis, but on my custom streaming machines, I like to keep them as minimal as possible and to that point would like to avoid installing pipewire+portals all together.

**Comment by ReenigneArcher** [comment]
The first option has an open PR here: https://github.com/LizardByte/Sunshine/pull/2507

Not sure what's left before that's ready to merge, but right now there are merge conflicts.

**Comment by Quackdoc** [comment]
Loosing kmsgrab also means loosing the capability of doing things such as session swapping with sunshine. Though this may be possible to bridge using something like gstreamer + pipewire, this won't be compatible with portals (and obviously) wont be compatible with wayland capture protocols either. 

I had realized recently that I do actually use this feature a small amount as depending on the task I may session swap to something such as gamescope or another x11 compositor. I think pipewire support would be able to adequately replace this, however im not sure if any tools for this currently exist that would adequately convert from kmsgrab to pipewire (gstreamer does not support kmscapture for instance)

AFAIK portals support also doesn't tend to lend itself well to unattended setups. As such I don't think portals is a suitable replacement for kmsgrab.

**Comment by Zamundaaa** [comment]
> AFAIK portals support also doesn't tend to lend itself well to unattended setups. As such I don't think portals is a suitable replacement for kmsgrab.

You may need to approve permissions for screen recording and remote input the very first time the app gets started, but it can restore those permissions later, without any input from the user.

> I have seen in the past other applications programmatically grab a specified screen without the need to interact with the Pipewire portal to select a screen or application.

There is no way around the portal, and no need to try to work around it either. With the remote desktop portal, after getting that initial permission, sunshine gets free access to both record all screens and to emulate input devices without any hacks, and without the user having to jump through hoops to set any of it up.

**Comment by Quackdoc** [comment]
> You may need to approve permissions for screen recording and remote input the very first time the app gets started, but it can restore those permissions later, without any input from the user.

Does KDE persist even across reboots? I know with gnome in the past even with `persist_mode=2`, the token can still expire after some time or after reboots. Though this may have changed, this has been a rather annoying issue for rustdesk users to contend with. 

Another issue with the portals is consistency, this is all assuming the portal even supports persist, which is itself not a given thanks to the nature of wayland. That's not to say portals isn't a good choice, for most of the setups I maintain, I have no doubt it would be the goto, and it's the only decent choice for supporting flatpak. I just don't see it as a proper replacement to kmsgrab.

**Comment by Lunaphied** [comment]
One approach that maybe should be explored is supporting something like https://github.com/nowrep/obs-vkcapture directly for Sunshine, which should be independent of compositor behavior and provides a path towards capturing HDR output. 

At least on KDE my experience is that portals are fine for say sharing something for a meeting but often there's issues like selecting a window capture when you needed to select a screen capture, applications launching to screens other than expected, screen capture not capturing certain games when doing screen capture and needing application capture or vis versa. It's too reliant on a lot of variables going right in an unattended environment even if token restoration was absolutely perfect (which it's not).

**Comment by Zamundaaa** [comment]
> Does KDE persist even across reboots? I know with gnome in the past even with persist_mode=2, the token can still expire after some time or after reboots. Though this may have changed, this has been a rather annoying issue for rustdesk users to contend with.

Of course it does. If that's not working correctly on Gnome, please report that bug to Gnome.

> Another issue with the portals is consistency

If something doesn't work, make a bug report. Every project trying to work around bugs with kmsgrab and root access just means that everything stays broken all the time, and everyone runs into problems like #93, #3298, #2955 and even more in the future. Kmsgrab isn't a proper replacement for using the actual APIs meant for what the project is trying to do.

Please, do not try to work around problems, it just creates more of them.

**Comment by Quackdoc** [comment]
Sway has merged support for the ext-image protocols https://github.com/swaywm/sway/pull/7976

**Comment by dElogics** [comment]
Pipewire is completely optional -- I'm not using it in anyway. What about systems which do not use it?

**Comment by nlutterman** [comment]
@dElogics, is continuing to use KMS grab a valid option for you?

**Comment by Quackdoc** [comment]
using x11 capture / wayland capture seems like the proper route forwards, but then something like arcan won't work. I think there should be alternatives to kmsgrab for sure, but kmsgrab for sure shouldn't be removed.

**Comment by dElogics** [comment]
> [@dElogics](https://github.com/dElogics), is continuing to use KMS grab a valid option for you? If it's not an option, rigging up waypipe may be an option if you're clever.

I'm using this to stream a complete desktop.

**Comment by Quackdoc** [comment]
cosmic now supports ext-image-copy-capture-v1

**Comment by njdom24** [comment]
The "wlroots" capture method now appears to be broken with Sway 1.11.

https://github.com/swaywm/sway/issues/8765

Edit: It works fine with `WLR_RENDERER=gles2` in 1.11, but Sway 1.10 capture worked fine with Vulkan. I assume it's more strict now.

**Comment by MwenDavo** [comment]
Has there been any development on this? This issue is the blocker for supporting HDR on NVidia systems since kmsgrab is not available, and NVFBC doesn't support HDR capture.

**Comment by Zamundaaa** [comment]
FTR I managed to put HDR screencasting together in KWin, so that issue will be going away in Plasma 6.6: https://invent.kde.org/plasma/kwin/-/merge_requests/8293

**Comment by TacoCake** [comment]
> FTR I managed to put HDR screencasting together in KWin, so that issue will be going away in Plasma 6.6: https://invent.kde.org/plasma/kwin/-/merge_requests/8293

For those would are waiting on this, it didn't get merged in time for 6.6, looks like the target milestone shifted to Plasma 6.7
